### PR TITLE
Initial implementation of a cactus stack.

### PR DESCRIPTION
**Pull request by ltratt**
An immutable cactus stuck (also called a spaghetti stack or parent pointer tree). A cactus stack is a (possibly empty) node with a (possibly null) pointer to a parent node. Pushing and popping return new nodes rather than mutating existing nodes.

```rust
use cactus::Cactus;
let c = Cactus::new();
assert!(c.is_empty());
let c2 = c.push(1);
assert_eq!(c2.len(), 1);
assert_eq!(*c2.val().unwrap(), 1);
let c3 = c2.pop().unwrap();
assert!(c3.is_empty());
```

From a given node one can create multiple sub-stacks:

```rust
use cactus::Cactus;
let c = Cactus::new().push(1);
let c2 = c.push(2);
let c3 = c.push(3);
assert!(c2 != c3);
assert_eq!(c2.iter().cloned().collect::<Vec<_>>(), [2, 1]);
assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [3, 1]);
```

Notice that the big difference between this and the (unfinished?) Kaktus library is that this one has immutable nodes. That means that we can use `Rc`s on their own without (expensive) `RefCell`s. And yes, it turns out that I do need a cactus stack for another program I'm writing!

### CODE REVIEW

**Comment by jaheba** [comment]
* I ran `cargo fmt` on the project and the few changes it made seem sensible.

* clippy suggest to remove the lifetimes here `pub fn iter<'a>(&'a self) -> CactusIter<'a, T> {`

**Comment by ltratt** [comment]
I've just done (and -- cheating -- force pushed) the Clippy changes.

File: src/lib.rs
Comment (jaheba) [inline src/lib.rs:101]: I think the whole method body can be replaced with `this.iter().count()`

**Comment by ltratt** [comment]
[My bad for the force push. Mea culpa.]

File: src/lib.rs
Comment (jaheba) [inline src/lib.rs]: This can be simplified as well

```rust
    pub fn push(&self, val: T) -> Cactus<T> {
        Cactus {
            node: Some(Rc::new(Node {
                val,
                parent: self.node.clone(),
            })),
        }
    }
```

File: src/lib.rs
Comment (jaheba) [inline src/lib.rs]: `self.node.as_ref().map(|node| Cactus { node: node.parent.clone() } )`

File: src/lib.rs
Comment (jaheba) [inline src/lib.rs]: just `self.node.as_ref()` should be sufficient

File: src/lib.rs
Comment (jaheba) [inline src/lib.rs]: just `self.node.as_ref()` should be sufficient

**Review by jaheba — CHANGES_REQUESTED** [review]

**Comment by jaheba** [comment]
Also, I personally prefer to not use one letter variable names, eg:

`self.node.as_ref().map(|n| &n.val)` -> `self.node.as_ref().map(|node| &node.val)`.

But everything else looks good!

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs:101]: I'm not sure I want to create the iterator just to throw it away, although maybe I'm over-optimising things...

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Agreed, that's definitely an improvement.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Agreed.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Not sure I understand.

Comment (ltratt) [reply src/lib.rs]: Ah, I'm with you.

**Comment by ltratt** [comment]
OK, I've pushed a commit which addresses most of your comments.

[I politely disagree about single-letter variables in closures. Once upon a time I liked long variable names. Then I realised that the best source code I was reading used shorter names, often single letter when in a tight scope. So since their source code was better than mine, I've switched... Occasionally I lapse into longer variable names but I try not to!]

File: src/lib.rs
Comment (jaheba) [reply src/lib.rs:101]: I would assume that it's optimised away and in my opinion the iterator version looks much cleaner.

Comment (ltratt) [reply src/lib.rs:101]: I did some (not very high quality) benchmarking. It looks like the iterator version might even be a bit quicker. It's certainly no slower. I've pushed an update accordingly.

**Comment by ltratt** [comment]
I think I'm done with Jasper's comments. Edd, do you have any? If not, I suggest that this is probably ready for squashing.

File: Cargo.toml
Comment (vext01) [inline Cargo.toml]: Does rust have the ability to specify a license in the Cargo file?

Comment (vext01) [reply Cargo.toml]: (random question)

File: src/lib.rs
Comment (vext01) [inline src/lib.rs:88]: I wonder if there are too many examples. E.g. it's pretty obvious how `is_empty` works.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: And the transitive parents, right?

File: src/lib.rs
Comment (vext01) [inline src/lib.rs:121]: I would have expected this method to have returned the item you popped and (because the stack is immutable) a new stack.

I guess it's trivial to get the current item off the top of the stack, but to call this `pop` may be confusing?

Not sure.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: This is not the fault of your code, but  `Option`'s `take()` confuses me. The docs are not helping much either.

```
Takes the value out of the option, leaving a None in its place.
```

But the `Option` is immutable, so how can it put a `None` in? What if the `Option` is already `None`?

File: src/lib.rs
Comment (vext01) [inline src/lib.rs:212]: OK, I don't know much about this stuff in Rust, but are you implementing a partial ordering? It seems to me that a stack should be totally ordered?

Maybe that's just the way you have to do it in rust.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: nice. (zip)

File: src/lib.rs
Comment (vext01) [inline src/lib.rs:243]: Would it be better to split this larger test into smaller ones for: e.g. `pop`, `is_empty`, like you did for `iter`?

I guess you could keep the big test too.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs:101]: Neat!

File: src/lib.rs
Comment (vext01) [inline src/lib.rs:71]: I was trying to figure out why you use an `Rc` here.

It's because the values in the nodes can come from different incompatible lifetimes?

I suspect if you were to clone `val` in push, the `Rc` could go? Not sure, and I'm pretty sure you have tried lots of variations and found this to be the only one you can get rust to accept (?).

Comment (vext01) [reply src/lib.rs:71]: I mean, in theory, because the stack is immutable, and since there is no sharing when a new stack is made on `push`, it seems to me that all the nodes should have the same lifetime and die together.

Comment (vext01) [reply src/lib.rs:71]: But then again, you never know with Rust.

**Review by vext01 — COMMENTED** [review]
Some comments, but mostly newbie questions.

You guys are much better at Rust than me.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs:71]: Because multiple children may point to the same parent so we can't guarantee to free a parent at the same point as a child.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: ?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs:121]: It could be called `parent` I suppose?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Notice that `fn take(&mut self)` i.e. `take` can mutate the `self` object.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs:212]: I'm not sure if/when one also needs `Eq` to be honest. Happy to be enlightened!

File: src/lib.rs
Comment (vext01) [reply src/lib.rs:71]: Hrm. I'm not going to persist on this, because it's likely I am wrong, but it seems like all of the nodes can die at the same time?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs:243]: I'd suggest this is best left as one test.

File: Cargo.toml
Comment (ltratt) [reply Cargo.toml]: Yes, I'm working on that stuff separately (probably for the next PR).

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs:71]: I'm not sure how to parse the "can" in that second sentence. Yes they *can* all die at the same time; but they don't *have* to die at the same time. If you think about it: I make a cactus C; push a child C'; I then let C' die; but C could/should be still alive at this point and I can push new children if I want.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: The new stack will contain the parent of the parent, and so on? I think I got confused by the wording. I would have worded it like:

```
Create a new cactus stack by cloning the existing one and adding `val` as the parent of the current `val`.
```

Or something like that.

Comment (ltratt) [reply src/lib.rs]: I agree that the wording here is tricky to get right and I'm all ears for improvements. However, this method doesn't clone the existing stack (it clones the existing stacks internal `Rc` cell, but that's an implementation detail that we want to shield the user from).

Comment (vext01) [reply src/lib.rs]: Doesn't it clone the `Option` containing the `Rc`? That being the case, this ties in with my lifetime question above. There can be sharing between stacks, which invalidates much of my argument, I think.

As for wording, can you see if you can capture the fact that the new stack contains all the old nodes and the one new one? I guess that it what I felt was missing.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs:121]: Your call. I don't want to come in guns blazing when I'm not really sure myself.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Yes, it clones the `Option` but that's purely because it's not worth doing `Some(node.unwrap().clone())` (i.e. it's just an optimisation of sorts).

I'll see if I can fix the wording.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: Right, so this is just a convenient way to unwrap an Option?

File: src/lib.rs
Comment (vext01) [reply src/lib.rs:212]: I'm not sure. The docs only confused me more.

I suggest you go ahead, and if this becomes an issue, fix it later.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: No, it moves the contents out. So if you had `Some(1)` then did `take`, you'd end up with a `None` in place of the `Some(1)`.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs:243]: ok

File: Cargo.toml
Comment (vext01) [reply Cargo.toml]: ok

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: ok

**Comment by ltratt** [comment]
I've renamed push and pop: see how you feel about the new versions.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs:71]: ok, i managed to convince myself this is probably ok.

**Comment by vext01** [comment]
At first I wasn't sure, but then I realised that since `pop` may not actually free up a node (someone else may have a reference to it), the notation you proposed actually is quite good.

What do you think though?

**Comment by ltratt** [comment]
Well, I *think* the new names are better at the moment. I might wake up tomorrow and change my mind of course ;)

**Comment by vext01** [comment]
It's a tough call. I'll let you choose.

**Comment by ltratt** [comment]
I think `child`/`parent` are an improvement and worth sticking with (unless I change my mind tomorrow at which point I can raise a new PR!).

@jaheba Thoughts?

If not, I think we're at squashing time...

File: src/lib.rs
Comment (jaheba) [reply src/lib.rs:212]: `PartialEq != PartialOrd`

PartialEq just means that a given value T does not have to be equal to itself (e.g. `NaN != NaN`). In contrast `Eq` signals that the type is also reflexive but doesn't add any other behaviour. If we implement `PartialEq` for our type, we also implement it for `Eq` so this should be fine.

**Comment by jaheba** [comment]
I personally prefer `push` and `pop`, since these are the operations on a stack. Also `child()` sounds like a getter for me. But it's your call.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs:212]: I'll take your word for it.

**Comment by ltratt** [comment]
The problem is that, for an immutable stack, `push` doesn't actually push anything (nor does `pop` pop anything), so I agree with Edd that they're confusing names. I'm not claiming the replacements are perfect though!

**Comment by ltratt** [comment]
OK, so unless anyone objects shortly, I'll squash this and propose it to be merged.

**Comment by ltratt** [comment]
OK, squashed, and (hopefully) ready for merging.

**Comment by vext01** [comment]
> ltratt: 5:54 PM @ebarrett btw, you can feel free to merge the cactus PR i think

Laurie is waiting on this, so let's push on.

**Comment by vext01** [comment]
All ready @ltratt?

**Comment by jaheba** [comment]
No objections here

**Comment by ltratt** [comment]
I think it's ready.

**Comment by ltratt** [comment]
@jaheba You'll need to approve the changes I think before this can be merged.

**Pull request merged**
